cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.resource;
+
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.ReconciliationLogger;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.QuorumInfo;
+
+/**
+ * Determines whether the given controller can be rolled without affecting the quorum.
+ */
+class KafkaQuorumCheck {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaQuorumCheck.class.getName());
+    private final Admin ac;
+    private final Reconciliation reconciliation;
+    private final Future<QuorumInfo> quorumInfoFuture;
+    private final long controllerQuorumFetchTimeoutMs;
+
+    KafkaQuorumCheck(Reconciliation reconciliation, Admin ac, long controllerQuorumFetchTimeoutMs) {
+        this.ac = ac;
+        this.reconciliation = reconciliation;
+        this.quorumInfoFuture = describeMetadataQuorum();
+        this.controllerQuorumFetchTimeoutMs = controllerQuorumFetchTimeoutMs;
+    }
+
+    /**
+     * Determine whether the given controller can be rolled without affecting the quorum
+     */
+    Future<Boolean> canRollController(int podId) {
+        LOGGER.debugCr(reconciliation, "Determining whether controller {} can be rolled", podId);
+        return quorumInfoFuture.map(info -> {
+            Map<Integer, Long> replicaIdToLastCaughtUpTimestampMap = info.voters().stream().collect(Collectors.toMap(
+                    QuorumInfo.ReplicaState::replicaId,
+                    replicaState -> replicaState.lastCaughtUpTimestamp().isPresent() ? replicaState.lastCaughtUpTimestamp().getAsLong() : -1));
+            boolean canRoll = isQuorumHealthy(info.leaderId(), replicaIdToLastCaughtUpTimestampMap);
+            if (!canRoll) {
+                LOGGER.debugCr(reconciliation, "Restart pod {} would affect the quorum", podId);
+            }
+            return canRoll;
+        }).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining whether it is safe to restart pod {}", podId, error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    private boolean isQuorumHealthy(int leaderId, Map<Integer, Long> replicaStates) {
+        int totalNumVoters = replicaStates.size();
+        AtomicInteger numOfCaughtUpVoters = new AtomicInteger();
+        long leaderLastCaughtUpTimestamp = replicaStates.get(leaderId);
+        if (leaderLastCaughtUpTimestamp < 0) {
+            LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for the leader replica {} ", leaderId);
+            return false;
```

Comment:
I see Kafka Java documentation stating this ...

> 
public [OptionalLong](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/OptionalLong.html) lastCaughtUpTimestamp()
Return the last millisecond timestamp at which this replica was known to be caught up with the leader.
Returns:
The value of the lastCaughtUpTime if known, empty otherwise

It seems that returning "empty" is possible, not sure when it could happen.

Reply:
I think this is possible, if the leader has just changed and follower has not started fetching from the new leader yet, either for a very short time or if there is a problem with the follower.

Reply:
But here we're talking about the leaders own state wrt _itself_. In other words, can the leader ever not be cause up to itself?

Reply:
That's actually true :-)

Reply:
I guess not. If we have a leaderId, then we probably would always have lastCaughtUpTimestamp so perhaps this check does not add any value.